File: backend/routes/blog_v2.py
"""Enterprise Blog V2 API routes."""

from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, HTTPException, Query, Request
from pydantic import BaseModel, Field

from routes.db import db, get_current_user
from services.blog_v2_service import (
    add_bookmark,
    ensure_blog_v2_seed,
    get_ai_summary,
    get_author_profile,
    get_bookmarks,
    get_history,
    get_home_payload,
    get_post_detail,
    get_recommendations,
    list_posts,
    remove_bookmark,
    write_history,
)
from services.blog_v2_engagement_service import (
    apply_owner_reminder_action,
    build_public_engagement_preview,
    claim_referral_bonus,
    consume_unlock_token_for_post,
    record_funnel_event,
    get_next_best_sequence,
    get_owner_engagement_loop,
    get_owner_rewards_snapshot,
    get_weekly_digest,
    send_weekly_digest_email,
    update_owner_reminder_settings,
)
from utils.access_control_engine import compute_effective_plan


router = APIRouter(prefix="/blog/v2", tags=["blog-v2"])


class BlogHistoryWriteRequest(BaseModel):
    post_id: str = Field(..., min_length=3)
    progress_percent: int = Field(default=0, ge=0, le=100)
    dwell_seconds: int = Field(default=0, ge=0, le=36000)


class BlogAISummaryRequest(BaseModel):
    slug: str = Field(..., min_length=2)


class BlogReminderSettingsRequest(BaseModel):
    enabled: Optional[bool] = None
    mode: Optional[str] = Field(default=None, description="adaptive | daily | three_per_week")
    digest_in_app_enabled: Optional[bool] = None
    digest_email_enabled: Optional[bool] = None


class BlogReminderActionRequest(BaseModel):
    action: str = Field(..., min_length=3, description="snooze_24h | dismiss | trigger_now")


class BlogReferralClaimRequest(BaseModel):
    invite_code: str = Field(..., min_length=4)


class BlogFunnelEventRequest(BaseModel):
    event_name: str = Field(..., min_length=3)
    context: Optional[dict] = None


async def _viewer_context(request: Request) -> tuple[Optional[object], Optional[str], str]:
    user = await get_current_user(request)
    if not user:
        return None, None, "free"

    user_doc = await db.users.find_one({"user_id": user.user_id}, {"_id": 0})
    if not user_doc:
        if getattr(user, "is_admin", False):
            return user, f"auth:{user.user_id}", "admin"
        return user, f"auth:{user.user_id}", "free"

    effective = compute_effective_plan(user_doc)
    if getattr(user, "is_admin", False) and effective != "admin":
        effective = "admin"
    return user, f"auth:{user.user_id}", effective


def _require_auth_owner(owner_id: Optional[str]) -> str:
    if not owner_id:
        raise HTTPException(status_code=401, detail="Authentication required")
    return owner_id


@router.get("/home")
async def blog_v2_home(request: Request):
    await ensure_blog_v2_seed(db)
    _, owner_id, viewer_plan = await _viewer_context(request)
    payload = await get_home_payload(db, viewer_plan=viewer_plan, owner_id=owner_id)
    payload["viewer_plan"] = viewer_plan
    return payload


@router.get("/posts")
async def blog_v2_posts(
    request: Request,
    q: str = Query(default=""),
    category: str = Query(default=""),
    tag: str = Query(default=""),
    author_slug: str = Query(default=""),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=12, ge=1, le=24),
    sort: str = Query(default="latest"),
):
    await ensure_blog_v2_seed(db)
    _, owner_id, viewer_plan = await _viewer_context(request)
    result = await list_posts(
        db,
        viewer_plan=viewer_plan,
        owner_id=owner_id,
        query=q,
        category=category,
        tag=tag,
        author_slug=author_slug,
        page=page,
        page_size=page_size,
        sort=sort,
    )
    result["viewer_plan"] = viewer_plan
    return result


@router.get("/posts/{slug}")
async def blog_v2_post_detail(request: Request, slug: str):
    await ensure_blog_v2_seed(db)
    _, owner_id, viewer_plan = await _viewer_context(request)
    detail = await get_post_detail(db, slug=slug, viewer_plan=viewer_plan, owner_id=owner_id)
    if not detail:
        raise HTTPException(status_code=404, detail="Post not found")
    detail["viewer_plan"] = viewer_plan
    return detail


@router.get("/authors/{author_slug}")
async def blog_v2_author(request: Request, author_slug: str):
    await ensure_blog_v2_seed(db)
    _, _, viewer_plan = await _viewer_context(request)
    payload = await get_author_profile(db, author_slug=author_slug, viewer_plan=viewer_plan)
    if not payload:
        raise HTTPException(status_code=404, detail="Author not found")
    payload["viewer_plan"] = viewer_plan
    return payload


@router.get("/search")
async def blog_v2_search(
    request: Request,
    q: str = Query(default="", min_length=1),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=12, ge=1, le=24),
):
    _, owner_id, viewer_plan = await _viewer_context(request)
    result = await list_posts(
        db,
        viewer_plan=viewer_plan,
        owner_id=owner_id,
        query=q,
        page=page,
        page_size=page_size,
        sort="latest",
    )
    result["query"] = q
    return result


@router.get("/recommendations")
async def blog_v2_recommendations(request: Request, limit: int = Query(default=8, ge=1, le=12)):
    _, owner_id, viewer_plan = await _viewer_context(request)
    picks = await get_recommendations(db, owner_id=owner_id, viewer_plan=viewer_plan, limit=limit)
    return {"items": picks, "viewer_plan": viewer_plan}


@router.post("/ai/summary")
async def blog_v2_ai_summary(request: Request, payload: BlogAISummaryRequest):
    _, owner_id, viewer_plan = await _viewer_context(request)
    summary = await get_ai_summary(db, slug=payload.slug, viewer_plan=viewer_plan, owner_id=owner_id)
    if not summary.get("ok"):
        raise HTTPException(status_code=404, detail="Post not found")
    return summary


@router.post("/me/history")
async def blog_v2_write_history(request: Request, payload: BlogHistoryWriteRequest):
    _, owner_id, _viewer_plan = await _viewer_context(request)
    owner = _require_auth_owner(owner_id)
    return await write_history(
        db,
        owner_id=owner,
        viewer_plan=_viewer_plan,
        post_id=payload.post_id,
        progress_percent=payload.progress_percent,
        dwell_seconds=payload.dwell_seconds,
    )


@router.get("/engagement-loop")
async def blog_v2_engagement_loop(request: Request):
    _, owner_id, viewer_plan = await _viewer_context(request)
    if not owner_id:
        return {"engagement_loop": build_public_engagement_preview(), "viewer_plan": viewer_plan}
    loop = await get_owner_engagement_loop(db, owner_id, viewer_plan)
    return {"engagement_loop": loop, "viewer_plan": viewer_plan}


@router.get("/next-best")
async def blog_v2_next_best(request: Request, limit: int = Query(default=5, ge=1, le=10), refresh: bool = Query(default=False)):
    _, owner_id, viewer_plan = await _viewer_context(request)
    owner = _require_auth_owner(owner_id)
    payload = await get_next_best_sequence(db, owner_id=owner, viewer_plan=viewer_plan, limit=limit, force_refresh=bool(refresh))
    return payload


@router.get("/weekly-digest")
async def blog_v2_weekly_digest(request: Request, refresh: bool = Query(default=False), send_email: bool = Query(default=False)):
    _, owner_id, viewer_plan = await _viewer_context(request)
    owner = _require_auth_owner(owner_id)
    digest = await get_weekly_digest(db, owner_id=owner, viewer_plan=viewer_plan, force_refresh=bool(refresh))
    email_result = None
    if bool(send_email):
        email_result = await send_weekly_digest_email(db, owner_id=owner, viewer_plan=viewer_plan)
    return {"digest": digest, "email_result": email_result}


@router.get("/me/rewards")
async def blog_v2_rewards_snapshot(request: Request):
    _, owner_id, viewer_plan = await _viewer_context(request)
    owner = _require_auth_owner(owner_id)
    payload = await get_owner_rewards_snapshot(db, owner_id=owner, viewer_plan=viewer_plan)
    return payload


@router.post("/me/referral-claim")
async def blog_v2_referral_claim(request: Request, payload: BlogReferralClaimRequest):
    _, owner_id, viewer_plan = await _viewer_context(request)
    owner = _require_auth_owner(owner_id)
    result = await claim_referral_bonus(db, claimer_owner_id=owner, viewer_plan=viewer_plan, invite_code=payload.invite_code)
    if not result.get("ok"):
        raise HTTPException(status_code=400, detail=result.get("reason") or "referral_claim_failed")
    await record_funnel_event(
        db,
        owner_id=owner,
        event_name="claim_success",
        context={"invite_code": payload.invite_code},
    )
    return result


@router.post("/me/unlock-premium/{post_id}")
async def blog_v2_unlock_premium(request: Request, post_id: str):
    _, owner_id, _viewer_plan = await _viewer_context(request)
    owner = _require_auth_owner(owner_id)
    result = await consume_unlock_token_for_post(db, owner_id=owner, post_id=post_id)
    if not result.get("ok"):
        raise HTTPException(status_code=400, detail=result.get("reason") or "unlock_failed")
    await record_funnel_event(
        db,
        owner_id=owner,
        event_name="premium_unlock_consumed",
        context={"post_id": post_id, "status": result.get("status")},
    )
    return result


@router.post("/me/funnel-event")
async def blog_v2_funnel_event(request: Request, payload: BlogFunnelEventRequest):
    _, owner_id, _viewer_plan = await _viewer_context(request)
    owner = _require_auth_owner(owner_id)
    result = await record_funnel_event(
        db,
        owner_id=owner,
        event_name=payload.event_name,
        context=payload.context or {},
    )
    if not result.get("ok"):
        raise HTTPException(status_code=400, detail=result.get("reason") or "funnel_event_rejected")
    return result


@router.patch("/me/reminder-settings")
async def blog_v2_update_reminder_settings(request: Request, payload: BlogReminderSettingsRequest):
    _, owner_id, viewer_plan = await _viewer_context(request)
    owner = _require_auth_owner(owner_id)
    loop = await update_owner_reminder_settings(
        db,
        owner_id=owner,
        viewer_plan=viewer_plan,
        enabled=payload.enabled,
        mode=payload.mode,
        digest_in_app_enabled=payload.digest_in_app_enabled,
        digest_email_enabled=payload.digest_email_enabled,
    )
    return {"ok": True, "engagement_loop": loop}


@router.post("/me/reminder-action")
async def blog_v2_reminder_action(request: Request, payload: BlogReminderActionRequest):
    _, owner_id, viewer_plan = await _viewer_context(request)
    owner = _require_auth_owner(owner_id)
    loop = await apply_owner_reminder_action(
        db,
        owner_id=owner,
        viewer_plan=viewer_plan,
        action=payload.action,
    )
    return {"ok": True, "engagement_loop": loop}


@router.get("/me/history")
async def blog_v2_history(request: Request):
    _, owner_id, viewer_plan = await _viewer_context(request)
    owner = _require_auth_owner(owner_id)
    rows = await get_history(db, owner_id=owner, viewer_plan=viewer_plan)
    return {"items": rows}


@router.get("/me/bookmarks")
async def blog_v2_bookmarks(request: Request):
    _, owner_id, viewer_plan = await _viewer_context(request)
    owner = _require_auth_owner(owner_id)
    rows = await get_bookmarks(db, owner_id=owner, viewer_plan=viewer_plan)
    return {"items": rows}


@router.post("/me/bookmarks/{post_id}")
async def blog_v2_add_bookmark(request: Request, post_id: str):
    _, owner_id, _viewer_plan = await _viewer_context(request)
    owner = _require_auth_owner(owner_id)
    result = await add_bookmark(db, owner_id=owner, post_id=post_id)
    if not result.get("ok"):
        raise HTTPException(status_code=404, detail="Post not found")
    return result


@router.delete("/me/bookmarks/{post_id}")
async def blog_v2_remove_bookmark(request: Request, post_id: str):
    _, owner_id, _viewer_plan = await _viewer_context(request)
    owner = _require_auth_owner(owner_id)
    return await remove_bookmark(db, owner_id=owner, post_id=post_id)

File: frontend/src/services/blogV2.ts
import api from './api';

export interface BlogCard {
  post_id: string;
  slug: string;
  title: string;
  excerpt: string;
  cover_image?: string;
  category: string;
  tags: string[];
  author_name: string;
  author_slug: string;
  author_role: string;
  published_at: string;
  date_display: string;
  read_time_minutes: number;
  premium_required: boolean;
  premium_locked: boolean;
  bookmarked: boolean;
  metrics?: {
    views?: number;
    bookmarks?: number;
    shares?: number;
  };
  history?: {
    progress_percent?: number;
    dwell_seconds?: number;
    updated_at?: string;
  };
}

export interface BlogAuthor {
  author_slug: string;
  name: string;
  role: string;
  bio: string;
  avatar_url?: string;
  focus_areas: string[];
  stats?: { posts?: number; total_views?: number };
  social?: { linkedin?: string; x?: string };
}

export interface BlogHomePayload {
  featured: BlogCard | null;
  trending: BlogCard[];
  latest: BlogCard[];
  categories: { name: string; count: number }[];
  authors: BlogAuthor[];
  stats: {
    total_posts: number;
    total_views: number;
    total_bookmarks: number;
    last_refreshed_at: string;
  };
  engagement_loop?: BlogEngagementLoop;
  viewer_plan: string;
}

export interface BlogEngagementLoop {
  streak: {
    current: number;
    longest: number;
    last_read_date: string;
    days_to_keep_streak: number;
    badges?: string[];
    next_milestone?: {
      threshold: number;
      remaining: number;
    } | null;
  };
  reminder: {
    enabled: boolean;
    mode: 'adaptive' | 'daily' | 'three_per_week' | string;
    status: string;
    message: string;
    next_reminder_at: string;
    snoozed_until: string;
    recommended_action: string;
    digest?: {
      in_app_enabled: boolean;
      email_enabled: boolean;
      last_digest_at: string;
      last_email_digest_at: string;
    };
  };
  conversion_trigger: {
    show_upgrade_nudge: boolean;
    reason: string;
    streak_threshold: number;
    locked_premium_available: boolean;
  };
  rewards?: {
    unlock_tokens: number;
    premium_preview_unlocks_used: number;
    referral_bonus_days_total: number;
    invite_code: string;
    invite_link: string;
    referrals_count: number;
  };
}

export interface BlogNextBestItem {
  post_id: string;
  slug: string;
  title: string;
  excerpt: string;
  cover_image?: string;
  category: string;
  author_name: string;
  read_time_minutes: number;
  premium_required: boolean;
  sequence_reason: string;
}

export interface BlogShareVariantOptimizerChannel {
  recommended_variant: 'short' | 'long' | 'benefit';
  recommended_label: string;
  winner_variant: 'short' | 'long' | 'benefit';
  winner_label: string;
  winner_share_pct: number;
  sample_size: number;
  strategy: string;
  exploration_rate: number;
  confidence_hint?: number;
  variant_metrics: Record<'short' | 'long' | 'benefit', {
    shares: number;
    unique_owners: number;
    surface_count: number;
    weighted_engagement: number;
    share_pct: number;
    draw: number;
  }>;
}

export interface BlogShareVariantOptimizer {
  window_days: number;
  generated_at: string;
  default_mode: 'auto';
  global_top_variant: 'short' | 'long' | 'benefit';
  global_top_variant_label: string;
  exploration_rate: number;
  channels: Record<string, BlogShareVariantOptimizerChannel>;
}

export interface BlogListPayload {
  items: BlogCard[];
  total: number;
  page: number;
  page_size: number;
  has_next: boolean;
  viewer_plan: string;
}

export interface BlogDetailPayload extends BlogCard {
  preview_blocks: { type: string; text: string }[];
  content_blocks: { type: string; text: string }[];
  premium_sections: { heading: string; content: string }[];
  premium_preview: { heading: string; content: string }[];
  premium_gate: {
    required: boolean;
    unlocked: boolean;
    cta: string;
  };
  author?: BlogAuthor;
  related_posts: BlogCard[];
  token_unlocked?: boolean;
  viewer_plan: string;
}

export interface BlogAuthorPayload {
  author: BlogAuthor;
  posts: BlogCard[];
  viewer_plan: string;
}

const normalizeError = (error: any) => {
  const status = error?.response?.status ?? error?.status ?? 500;
  const message =
    error?.response?.data?.detail ||
    error?.response?.data?.message ||
    error?.message ||
    'Request failed';
  return { status, message };
};

export const isUnauthorized = (error: any) => normalizeError(error).status === 401;

export const getBlogHome = async (): Promise<BlogHomePayload> => {
  const response = await api.get('/blog/v2/home');
  return response.data as BlogHomePayload;
};

export const getBlogPosts = async (params: {
  q?: string;
  category?: string;
  tag?: string;
  author_slug?: string;
  page?: number;
  page_size?: number;
  sort?: string;
}): Promise<BlogListPayload> => {
  const response = await api.get('/blog/v2/posts', { params });
  return response.data as BlogListPayload;
};

export const searchBlogPosts = async (q: string, page = 1, page_size = 12): Promise<BlogListPayload> => {
  const response = await api.get('/blog/v2/search', { params: { q, page, page_size } });
  return response.data as BlogListPayload;
};

export const getBlogPost = async (slug: string): Promise<BlogDetailPayload> => {
  const response = await api.get(`/blog/v2/posts/${encodeURIComponent(slug)}`);
  return response.data as BlogDetailPayload;
};

export const getBlogAuthor = async (authorSlug: string): Promise<BlogAuthorPayload> => {
  const response = await api.get(`/blog/v2/authors/${encodeURIComponent(authorSlug)}`);
  return response.data as BlogAuthorPayload;
};

export const getBlogRecommendations = async (limit = 8): Promise<{ items: BlogCard[]; viewer_plan: string }> => {
  const response = await api.get('/blog/v2/recommendations', { params: { limit } });
  return response.data;
};

export const getAiSummary = async (slug: string): Promise<{
  ok: boolean;
  summary: string;
  key_takeaways: string[];
  action_plan: string[];
  model: string;
  cached: boolean;
  upgrade_required?: boolean;
}> => {
  const response = await api.post('/blog/v2/ai/summary', { slug });
  return response.data;
};

export const writeBlogHistory = async (post_id: string, progress_percent: number, dwell_seconds: number) => {
  const response = await api.post('/blog/v2/me/history', {
    post_id,
    progress_percent,
    dwell_seconds,
  });
  return response.data;
};

export const getBlogEngagementLoop = async (): Promise<{ engagement_loop: BlogEngagementLoop; viewer_plan: string }> => {
  const response = await api.get('/blog/v2/engagement-loop');
  return response.data;
};

export const updateBlogReminderSettings = async (payload: {
  enabled?: boolean;
  mode?: 'adaptive' | 'daily' | 'three_per_week';
  digest_in_app_enabled?: boolean;
  digest_email_enabled?: boolean;
}) => {
  const response = await api.patch('/blog/v2/me/reminder-settings', payload);
  return response.data as { ok: boolean; engagement_loop: BlogEngagementLoop };
};

export const sendBlogReminderAction = async (action: 'snooze_24h' | 'dismiss' | 'trigger_now') => {
  const response = await api.post('/blog/v2/me/reminder-action', { action });
  return response.data as { ok: boolean; engagement_loop: BlogEngagementLoop };
};

export const getBlogNextBest = async (limit = 5, refresh = false): Promise<{
  week_key: string;
  strategy: string;
  source: string;
  generated_at: string;
  items: BlogNextBestItem[];
}> => {
  const response = await api.get('/blog/v2/next-best', { params: { limit, refresh } });
  return response.data;
};

export const getBlogWeeklyDigest = async (refresh = false, send_email = false): Promise<{
  digest: {
    week_key: string;
    generated_at: string;
    headline: string;
    highlights: string[];
    email_sent_at: string;
    next_best: {
      items: BlogNextBestItem[];
    };
  };
  email_result?: {
    ok: boolean;
    status: string;
    reason?: string;
  } | null;
}> => {
  const response = await api.get('/blog/v2/weekly-digest', { params: { refresh, send_email } });
  return response.data;
};

export const getBlogRewards = async (): Promise<{
  streak: BlogEngagementLoop['streak'];
  rewards: NonNullable<BlogEngagementLoop['rewards']>;
  conversion_trigger: BlogEngagementLoop['conversion_trigger'];
  unlockables: {
    premium_preview_tokens: number;
    available: boolean;
  };
  referral: {
    invite_code: string;
    invite_link: string;
    referrals_count: number;
    bonus_days_total: number;
  };
  weekly_digest: {
    week_key: string;
    generated_at: string;
    email_sent_at: string;
    headline: string;
  };
  next_best: {
    week_key: string;
    generated_at: string;
    strategy: string;
    items: BlogNextBestItem[];
  };
  channel_attribution?: {
    window_days: number;
    top_channel: string;
    totals: {
      clicks: number;
      unique_clicks: number;
      claim_attributed_clicks: number;
    };
    channels: Record<string, {
      clicks: number;
      unique_clicks: number;
      claim_attributed_clicks: number;
    }>;
    trend: { date: string; clicks: number }[];
  };
  share_variant_optimizer?: BlogShareVariantOptimizer;
}> => {
  const response = await api.get('/blog/v2/me/rewards');
  return response.data;
};

export const claimBlogReferralBonus = async (invite_code: string): Promise<{
  ok: boolean;
  boost_days: number;
  reason: string;
  rewards: ReturnType<typeof getBlogRewards> extends Promise<infer T> ? T : never;
}> => {
  const response = await api.post('/blog/v2/me/referral-claim', { invite_code });
  return response.data;
};

export const unlockBlogPremiumWithToken = async (post_id: string): Promise<{
  ok: boolean;
  status: 'unlocked' | 'already_unlocked';
  post_id: string;
  remaining_tokens: number;
}> => {
  const response = await api.post(`/blog/v2/me/unlock-premium/${encodeURIComponent(post_id)}`);
  return response.data;
};

export const trackBlogFunnelEvent = async (event_name: string, context: Record<string, any> = {}) => {
  const response = await api.post('/blog/v2/me/funnel-event', { event_name, context });
  return response.data;
};

export const getBlogHistory = async (): Promise<{ items: BlogCard[] }> => {
  const response = await api.get('/blog/v2/me/history');
  return response.data;
};

export const getBlogBookmarks = async (): Promise<{ items: BlogCard[] }> => {
  const response = await api.get('/blog/v2/me/bookmarks');
  return response.data;
};

export const addBlogBookmark = async (postId: string) => {
  const response = await api.post(`/blog/v2/me/bookmarks/${encodeURIComponent(postId)}`);
  return response.data;
};

export const removeBlogBookmark = async (postId: string) => {
  const response = await api.delete(`/blog/v2/me/bookmarks/${encodeURIComponent(postId)}`);
  return response.data;
};

export const parseBlogError = normalizeError;
